python/trace-python.c

```diff
@@ -1017,6 +1022,14 @@ static void __attribute__((destructor)) uftrace_trace_python_finish(void)
 	free(main_dir);
 }
 
+static PyObject *uftrace_trace_python_exit(PyObject *self, PyObject *obj)
+{
+	int n = PyLong_AsLong(obj);
```

Comment:
I think it's better to emit an entry for "os._exit" so that users can know it's exited by that.

Reply:
Hmm.. I dont’t get what you mean. Could you explain what you mean by “emit an entry”?

Reply:
Something like this

```
__main__.<module>() {
   a() {
     b() {
       c() {
         posix.getpid();
       } /* c */
     } /* b */
   } /* a */
  os._exit() {          <--- here
```

Of course it'd also show remaining functions, but I think that's clearer.

```
uftrace stopped tracing with remaining functions
================================================
task: 1234
[1] os._exit
[0] __main__.<module>
```

Reply:
Nevermind, I can add that later.

Reply:
It actually captures the entry of `os_exit`, but didn't fully write the record. I can see that with the following change.
```diff
diff --git a/python/trace-python.c b/python/trace-python.c
index 84ffb873..9ddd003f 100644
--- a/python/trace-python.c
+++ b/python/trace-python.c
@@ -981,14 +981,18 @@ static PyObject *uftrace_trace_python(PyObject *self, PyObject *args)
                Py_RETURN_NONE;
 
        if (!strcmp(event, "call") || !strcmp(event, "c_call")) {
-               if (can_trace(true, sym))
+               if (can_trace(true, sym)) {
+                       fprintf(stderr, "[ call ] sym->name = %s\n", sym->name);
                        cygprof_enter(sym->addr, 0);
+               }
                else
                        Py_RETURN_NONE;
        }
        else if (!strcmp(event, "return") || !strcmp(event, "c_return")) {
-               if (can_trace(false, sym))
+               if (can_trace(false, sym)) {
+                       fprintf(stderr, "[return] sym->name = %s\n", sym->name);
                        cygprof_exit(0, 0);
+               }
                else
                        Py_RETURN_NONE;
        }
```
It shows the `os_exit` and `uftrace_python.exit` as well.
```
$ uftrace tests/s-abc-exit.py 
[ call ] sym->name = __main__.<module>
[ call ] sym->name = a
[ call ] sym->name = b
[ call ] sym->name = c
[ call ] sym->name = posix.getpid
[return] sym->name = posix.getpid
[return] sym->name = c
[return] sym->name = b
[return] sym->name = a
[ call ] sym->name = os_exit
[ call ] sym->name = uftrace_python.exit
# DURATION     TID     FUNCTION
            [ 12145] | __main__.<module>() {
            [ 12145] |   a() {
            [ 12145] |     b() {
            [ 12145] |       c() {
   2.254 us [ 12145] |         posix.getpid();
   8.856 us [ 12145] |       } /* c */
  14.264 us [ 12145] |     } /* b */
  20.208 us [ 12145] |   } /* a */
            [ 12145] |   /* linux:task-exit */

uftrace stopped tracing with remaining functions
================================================
task: 12145
[0] __main__.<module>
```
I'm not sure why it's not fully written in uftrace record.

Reply:
Maybe it didn't flush the rstack until it sees a exit record.